health/health.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package health
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/projectcalico/felix/set"
+)
+
+type HealthSource uint8
+
+const (
+	CALC_GRAPH_IN_SYNC HealthSource = iota
+	CALC_GRAPH_RUNNING
+)
+
+type HealthIndicator struct {
+	// The source of this health indicator.
+	Source HealthSource
+
+	// How long the indicator is valid for.  In other words, if it continues operating normally,
+	// the source expects to refresh this indicator before this timeout.
+	Timeout time.Duration
+}
+
+func ServeHealth(port int, neededForReady set.Set, neededForLive set.Set, c <-chan HealthIndicator) {
+	ready := false
+	live := false
+	mutex := &sync.Mutex{}
+
+	go func() {
+		currentHealth := set.New()
+		timer := map[HealthSource]*time.Timer{}
+		timeoutC := map[HealthSource]<-chan time.Time{}
+		timeoutC[CALC_GRAPH_IN_SYNC] = nil
+		timeoutC[CALC_GRAPH_RUNNING] = nil
+
+		indicatorCurrent := func(source HealthSource, timeout time.Duration) {
+			log.WithField("source", source).Debug("Health indicator current")
+			currentHealth.Add(source)
+			if timer[source] != nil {
+				timer[source].Stop()
+			}
+			timer[source] = time.NewTimer(timeout)
+			timeoutC[source] = timer[source].C
+		}
+		indicatorExpired := func(source HealthSource) {
+			log.WithField("source", source).Debug("Health indicator expired")
+			currentHealth.Discard(source)
+			timeoutC[source] = nil
+		}
+
+		for {
+			select {
+			case indicator := <-c:
+				indicatorCurrent(indicator.Source, indicator.Timeout)
+			case <-timeoutC[CALC_GRAPH_IN_SYNC]:
+				indicatorExpired(CALC_GRAPH_IN_SYNC)
+			case <-timeoutC[CALC_GRAPH_RUNNING]:
+				indicatorExpired(CALC_GRAPH_RUNNING)
+			}
+			mutex.Lock()
+			ready = currentHealth.ContainsAll(neededForReady)
+			live = currentHealth.ContainsAll(neededForLive)
+			log.WithFields(log.Fields{
+				"ready": ready,
+				"live":  live,
+			}).Debug("Health now")
+			mutex.Unlock()
+		}
+	}()
+
+	for {
+		log.WithField("port", port).Info("Starting health endpoints")
+		http.HandleFunc("/readiness", func(rsp http.ResponseWriter, req *http.Request) {
```

Comment:
Do these need to be reregistered each time around the loop?  Are there any side-effects of doing that?